src/progressbar.rs

```diff
@@ -0,0 +1,432 @@
+            progress_bar: RwLock::new(indicatif_pb),
+            multiline: RwLock::new(Some(MultiLineProgressState::new(true))),
+            kind: ProgressBarKind::Progress.into(),
+            title: RwLock::new(None),
+        });
+
+        for bar in self.progress_bars.write().unwrap().iter_mut() {
+            bar.multiline
+                .write()
+                .unwrap()
+                .as_mut()
+                .unwrap()
+                .set_is_last(false);
+            bar.format_as_progressbar_multi(&*theme);
+        }
+        pb.set_style(theme.multiprogress_template(&pb, ThemeState::Active));
+        self.progress_bars.write().unwrap().push(pb.clone());
+
+        ProgressBarWrapper::new(pb.clone(), self)
+    }
+
+    pub fn add_downloadbar(&self) -> ProgressBarWrapper {
+        let theme = THEME.lock().unwrap();
+
+        let indicatif_pb = self.inner.add(indicatif::ProgressBar::new(100));
+
+        let pb = Arc::new(ProgressBar {
+            progress_bar: RwLock::new(indicatif_pb),
+            multiline: RwLock::new(Some(MultiLineProgressState::new(true))),
+            kind: ProgressBarKind::Download.into(),
+            title: RwLock::new(None),
+        });
+
+        for bar in self.progress_bars.read().unwrap().iter() {
+            bar.multiline
+                .write()
+                .unwrap()
+                .as_mut()
+                .unwrap()
+                .set_is_last(false);
+            bar.format_as_progressbar_multi(&*theme);
+        }
+        pb.set_style(theme.multiprogress_template(&pb, ThemeState::Active));
+        self.progress_bars.write().unwrap().push(pb.clone());
+
+        ProgressBarWrapper::new(pb.clone(), self)
+    }
+
+    pub fn add_spinner(&self) -> ProgressBarWrapper {
+        let indicatif_pb = self.inner.add(indicatif::ProgressBar::new_spinner());
+
+        let pb = Arc::new(ProgressBar {
+            progress_bar: RwLock::new(indicatif_pb),
+            multiline: RwLock::new(None),
+            kind: ProgressBarKind::Progress.into(),
+            title: RwLock::new(None),
+        });
+
+        self.progress_bars.write().unwrap().push(pb.clone());
+
+        ProgressBarWrapper::new(pb.clone(), self)
+    }
+
+    fn check_finish(&self, pb: Arc<ProgressBar>) {
+        if self.is_finished() {
+            let bars = self.progress_bars.write().unwrap();
+
+            for _ in 0..=bars.len() {
+                Term::stderr().move_cursor_up(1).unwrap();
+                Term::stderr().clear_line().unwrap();
+            }
+
+            let theme = THEME.lock().unwrap();
+            let titles = bars
+                .iter()
+                .map(|bar| bar.title.read().unwrap().as_ref().unwrap().to_string())
+                .collect::<Vec<String>>();
+
+            let title_refs = titles.iter().map(AsRef::as_ref).collect::<Vec<&str>>();
+            term_write(theme.format_multiprogres_stop(&self.heading, &title_refs)).unwrap();
```

Comment:
Wow, this is really dirty... I understand it works though. Maybe, [MultiProgress](https://docs.rs/indicatif/latest/indicatif/struct.MultiProgress.html) could be utilized somehow, I see there are customizations and `println`...

Reply:
Yeah, it's not pretty, but it is the most effective way I've found so far to handle this. We're already using `MultiProgress` here, but it has no idea about `S_BAR` etc and overwrites it, thus removing `cliclack` styling.

Atm my focus is on getting things to work for a specific CLI in [stacks](https://github.com/stacks-network) -- I'm more than willing to make an effort to find a cleaner way to do this, but I'm afraid I'm a little short on bandwidth to spend too much more time on this now :cry:

Reply:
Btw a lot of the "ugliness" does come from the usage of `RwLock`.  While the example doesn't provide a real-world-scenario, in reality a multi-progressbar is most likely being updated from multiple threads. This can be forced onto the consumer, of course, but it makes the usage of `cliclack` much simpler in multi-threaded scenarios to not have to worry about this.